### PR TITLE
Move Catalan Section

### PR DESCRIPTION
**Pull request by brebenelmihnea**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: content/6_Advanced/Catalan.mdx [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+	auto fact = [&](int x) -> int {
+		int ans = 1;
+		for (int i = 1; i <= x; i++) { ans = 1LL * ans * i % MOD; }
+		return ans;
+	};
+
+	auto comb = [&](int n, int k) {
+		return 1LL * fact(n) * inv(fact(k)) % MOD * inv(fact(n - k)) % MOD;
+	};
```
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx]: done

File: content/6_Advanced/Catalan.mdx [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+```cpp
+#include <iostream>
+
+using namespace std;
+
+int main() {
+	const int MOD = 1e9 + 7;
+
+	int N;
+	cin >> N;
```
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx]: done

File: content/6_Advanced/Catalan.mdx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+
+using namespace std;
+
+int main() {
+	const int MOD = 1e9 + 7;
```
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx]: done

File: content/6_Advanced/Catalan.mdx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+	int N;
+	cin >> N;
+
+	if (N & 1) {
```
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx]: done

File: content/6_Advanced/Catalan.mdx [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+using namespace std;
+
+const int MOD = 1e9 + 7;
+const int MAXN = 1e6;
+
+long long fac[MAXN + 1];
+long long inv[MAXN + 1];
+
+/** @return x^n modulo m in O(log p) time. */
+long long exp(long long x, long long n, long long m) {
```
Comment (SansPapyrus683) [inline content/6_Advanced/Catalan.mdx:112]: wrap this stuff and the other stuff in a codesnip
call it uh
```
Combinatorics Functions (from the module)
```

**Comment by SansPapyrus683** [comment]
btw since this is new module i'm not merging until benq approves it

File: content/6_Advanced/Catalan.mdx (hunk 5 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx:113]: alright done

File: content/6_Advanced/Catalan.mdx [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+	if (n % 2) {
+		cout << 0;
+		return 0;
+	}
+
+	factorial(MOD);
+	inverses(MOD);
+
+	cout << exp(n / 2 + 1, MOD - 2, MOD) * choose(n, n / 2, MOD) % MOD;
```
Comment (SansPapyrus683) [inline content/6_Advanced/Catalan.mdx:157]: endl here

File: content/6_Advanced/Catalan.mdx [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+	return fac[n] * inv[r] % p * inv[n - r] % p;
+}
+// EndCodeSnip
+
+int main() {
+	long long n;
+	cin >> n;
+
+	if (n % 2) {
+		cout << 0;
```
Comment (SansPapyrus683) [inline content/6_Advanced/Catalan.mdx:150]: here too

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/6_Advanced/Catalan.mdx (hunk 6 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx:157]: yep

File: content/6_Advanced/Catalan.mdx (hunk 7 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx:150]: yep

File: content/6_Advanced/Catalan.mdx [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+		title="Catalan Numbers"
+		url="https://cp-algorithms.com/combinatorics/catalan-numbers.html"
+		starred
+	>
+		Well documented article.
+	</Resource>
+</Resources>
+
+The [Catalan numbers](https://en.wikipedia.org/wiki/Catalan_number) are a sequence of positive integers that can be very useful in counting problems in combinatorics.
+The $n$-th Catalan can be expressed as following using binomial coeffiecients:
```
Comment (bqi343) [inline content/6_Advanced/Catalan.mdx:23]:
```suggestion
The $n$-th Catalan can be expressed as follows using binomial coefficients:
```

File: content/6_Advanced/Catalan.mdx [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+It's also equal to the number of full binary trees with $n+1$ leaves.
+The following image shows the $5$ binary trees with $4$ leaves:
+
+![binary-trees](./assets/catalan-binary-trees.png)
+
+$C_n$ is also the number of monotonic lattice paths along the edges of a $n \times n$ grid that don't pass above the diagonal.
+The paths start in the lower left corner and ends in the upper right corner.
+
+For example, there are $C_4=14$ paths in a $4 \times 4 grid$:
```
Comment (bqi343) [inline content/6_Advanced/Catalan.mdx:68]: grid should not be in `$...$`

File: content/6_Advanced/Catalan.mdx [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+<FocusProblem problem="sample" />
+
+## Explanation
+
+The problem is a direct application of Catalan numbers. The answer for $N$ is the $N/2$ Catalan Number.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N \cdot \log{N})$
```
Comment (bqi343) [inline content/6_Advanced/Catalan.mdx:95]: isn't it $O(N)$?

File: content/4_Gold/Combinatorics.problems.json [hunk 11]
```diff
@@ -30,21 +30,6 @@
       }
     }
   ],
-  "sample3": [
-    {
-      "uniqueId": "skyline",
```
Comment (bqi343) [inline content/4_Gold/Combinatorics.problems.json:35]: are you not going to add this problem to the new module?

File: content/6_Advanced/Catalan.mdx [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+The following image shows the $5$ binary trees with $4$ leaves:
+
+![binary-trees](./assets/catalan-binary-trees.png)
+
+$C_n$ is also the number of monotonic lattice paths along the edges of a $n \times n$ grid that don't pass above the diagonal.
+The paths start in the lower left corner and ends in the upper right corner.
+
+For example, there are $C_4=14$ paths in a $4 \times 4 grid$:
+
+![lattice points](./assets/catalan-lattice-paths.png)
```
Comment (ryanchou-dev) [inline content/6_Advanced/Catalan.mdx:70]: i think there was a github issue/comment about this, but the bottom left image is wrong

could you fix it? or i can do some image editing lol

File: content/6_Advanced/Catalan.mdx [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+	factorial(MOD);
+	inverses(MOD);
+
+	cout << exp(n / 2 + 1, MOD - 2, MOD) * choose(n, n / 2, MOD) % MOD << endl;
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
```
Comment (ryanchou-dev) [inline content/6_Advanced/Catalan.mdx:163]: add problems header

**Review by ryanchou-dev — CHANGES_REQUESTED** [review]

File: content/6_Advanced/Catalan.mdx (hunk 8 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx:23]: ok

File: content/6_Advanced/Catalan.mdx (hunk 10 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx:95]: ok

File: content/6_Advanced/Catalan.mdx (hunk 12 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx:70]: Solved
image

File: content/6_Advanced/Catalan.mdx (hunk 13 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx:163]: ok

File: content/4_Gold/Combinatorics.problems.json (hunk 11 above)
Comment (brebenelmihnea) [reply content/4_Gold/Combinatorics.problems.json:35]: Well I really don't know how to demonstrate why catalan number is the answer.

File: content/6_Advanced/Catalan.mdx (hunk 9 above)
Comment (brebenelmihnea) [reply content/6_Advanced/Catalan.mdx:68]: yep

File: content/4_Gold/Combinatorics.problems.json (hunk 11 above)
Comment (brebenelmihnea) [reply content/4_Gold/Combinatorics.problems.json:35]: Alright I added it as a practice problem.

**Review by bqi343 — APPROVED** [review]

**Review by SansPapyrus683 — APPROVED** [review]
seems legit

**Review by ryanchou-dev — APPROVED** [review]

**Pull request merged**